Move in-tree kernel modules to separate derivation output

jmbaur opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

The kernel image takes up a significant amount of space in the derivation where kernel modules are installed (even more so on aarch64 where we install the uncompressed `Image` file). Since the kernel image is not functionally needed in this derivation (by the time kernel modules would be of any use, we've already booted into our kernel), we can save some space by splitting out the modules into their own output. This change doesn't really do a lot for pure NixOS closures today, since the kernel image is still in the closure of `config.system.build.toplevel`, however the linux kernel build recipe in nixpkgs is super useful for more than just NixOS use-cases anyways. Plus, it's not out of the realm of possibility for NixOS's `toplevel` to remove the reference to the kernel image, which would be especially helpful in saving space on image-based systems where the ESP and/or boot partition is built and populated ahead of time and not part of the switch-to-configuration/bootloader-installation step.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

K900 commented:
CC @emilazy

alyssais reviewed (approved):
Seems sensible enough to me but haven't tested.  I guess there's not anything we can really do to make it backwards compatible without defeating the purpose by adding a reference…

emilazy commented:
Yeah I think it makes sense to batch this with my other target change since that will be (deliberately) breaking.

> That would be great! Would be amazing to build as many kernel image types for a given architecture as possible (like the ideas in #324419), to allow for the user to pick the one they want later without having to incur a rebuild.

Right, yeah, that’s the idea, especially as it means we can more safely prefer compressed UEFI stub images on AArch64. Though `uImage` I am dropping entirely.

jmbaur commented:
This is ready. I've tested locally with the `systemd-initrd-modprobe` test, though ofborg seems to be chugging on it.

emilazy commented:
We can avoid contorting the output name by just unsetting `modules` and using `${outputs[modules]}`, I believe. Might require `__structuredAttrs` but would be as good a time as any to migrate to that if we're not already on it anyway.

jmbaur commented:
> We can avoid contorting the output name by just unsetting `modules` and using `${outputs[modules]}`, I believe. Might require `__structuredAttrs` but would be as good a time as any to migrate to that if we're not already on it anyway.

Should be good now!

emilazy commented on pkgs/os-specific/linux/kernel/manual-config.nix:
```diff
@@ -439,10 +449,14 @@ lib.makeOverridable (
             installFlagsArray+=("INSTALL_MOD_STRIP=1")
           fi
+
+          modulesOutput=$(jq --raw-output '.outputs.modules' $NIX_ATTRS_JSON_FILE)
```
Is this necessary? `${outputs[modules]}` should work directly when `__structuredAttrs` is set. I would rather avoid the `jq` dependency here if we already get it natively.

emilazy commented on pkgs/os-specific/linux/kernel/manual-config.nix:
```diff
@@ -337,15 +342,20 @@ lib.makeOverridable (
+          [ "INSTALL_PATH=${placeholder "out"}" ]
           ++ optionals buildDTBs [
             "dtbs_install"
             "INSTALL_DTBS_PATH=$(out)/dtbs"
```
This is now incorrect. (Could you split up the switch to `__structuredAttrs` into another commit?)

jmbaur replied:
How is it incorrect? Does `placeholder` not work with `__structuredAttrs`? Regardless, the change to `placeholder` here is on a piece of code technically not handled by the intended change of my PR, so I will change back to the `$(out)` style.

emilazy commented:
BTW, this should be release noted as a breaking change.

emilazy commented on pkgs/os-specific/linux/kernel/manual-config.nix:
```diff
@@ -439,10 +449,13 @@ lib.makeOverridable (
             installFlagsArray+=("INSTALL_MOD_STRIP=1")
           fi
+
+          installFlagsArray+=("INSTALL_MOD_PATH=''${outputs[modules]}")
```
This could just be `INSTALL_MOD_PATH=${placeholder "modules"}` in the flags, right? Am I misunderstanding?

jmbaur commented:
@emilazy I ran into a bunch of stdenv hooks issues with __structuredAttrs on this one. Since `placeholder` works, I propose we stick to keeping  __structuredAttrs out of this change and introduce it at some later point. Let me know what you think about the current set of changes.

emilazy reviewed (commented):
Sorry for the long delay in reviewing; I’ve been busy lately.

I found the time to investigate what’s going on here as part of my further work on splitting up kernel outputs. It turns out that the Linux build system doesn’t care about the `modules` environment variable, but that it instead triggers a confusing behaviour in Make.

GitHub suggestions don’t make it convenient to show changes in the rest of the file, so here’s my proposed diff on top of your changes, including an explanation of the weird errors we were getting:

```diff
diff --git a/pkgs/os-specific/linux/kernel/manual-config.nix b/pkgs/os-specific/linux/kernel/manual-config.nix
index 7f22606c83..eb00132f26 100644
--- a/pkgs/os-specific/linux/kernel/manual-config.nix
+++ b/pkgs/os-specific/linux/kernel/manual-config.nix
@@ -183,6 +183,8 @@
         ];
       })
       // {
+        __structuredAttrs = true;
+
         passthru = rec {
           inherit
             version
@@ -309,17 +311,17 @@
 
           # reads the existing .config file and prompts the user for options in
           # the current kernel source that are not found in the file.
-          make $makeFlags "''${makeFlagsArray[@]}" oldconfig
+          make "''${makeFlags[@]}" oldconfig
           runHook postConfigure
 
-          make $makeFlags "''${makeFlagsArray[@]}" prepare
+          make "''${makeFlags[@]}" prepare
           actualModDirVersion="$(cat $buildRoot/include/config/kernel.release)"
           if [ "$actualModDirVersion" != "${modDirVersion}" ]; then
             echo "Error: modDirVersion ${modDirVersion} specified in the Nix expression is wrong, it should be: $actualModDirVersion"
             exit 1
           fi
 
-          buildFlagsArray+=("KBUILD_BUILD_TIMESTAMP=$(date -u -d @$SOURCE_DATE_EPOCH)")
+          buildFlags+=("KBUILD_BUILD_TIMESTAMP=$(date -u -d @$SOURCE_DATE_EPOCH)")
 
           cd $buildRoot
         '';
@@ -345,20 +347,6 @@
             "INSTALL_DTBS_PATH=${placeholder "out"}/dtbs"
           ];
 
-        # We have a derivation output named "modules", but the kernel build
-        # system consumes an environment variable named "modules", which
-        # means a different thing, so we unset our definition here and
-        # consume it later via __structuredAttrs.
-        preBuild = lib.optionalString isModular ''
-          unset modules
-        '';
-
-        # Some fixup hooks expect environment variables to be set for all
-        # outputs, so we undo what we did in preBuild.
-        preFixup = lib.optionalString isModular ''
-          modules=${placeholder "modules"}
-        '';
-
         preInstall =
           let
             # All we really need to do here is copy the final image and System.map to $out,
@@ -413,7 +401,7 @@
             '';
           in
           ''
-            installFlagsArray+=("-j$NIX_BUILD_CORES")
+            installFlags+=("-j$NIX_BUILD_CORES")
             export HOME=${installkernel}
           '';
 
@@ -450,13 +438,11 @@
           mkdir -p $dev
           cp vmlinux $dev/
           if [ -z "''${dontStrip-}" ]; then
-            installFlagsArray+=("INSTALL_MOD_STRIP=1")
+            installFlags+=("INSTALL_MOD_STRIP=1")
           fi
 
-          make modules_install $makeFlags "''${makeFlagsArray[@]}" \
-            $installFlags "''${installFlagsArray[@]}"
-          unlink ${placeholder "modules"}/lib/modules/${modDirVersion}/build
-          rm -f ${placeholder "modules"}/lib/modules/${modDirVersion}/source
+          make modules_install "''${makeFlags[@]}" "''${installFlags[@]}"
+          unlink $modules/lib/modules/${modDirVersion}/build
 
           mkdir -p $dev/lib/modules/${modDirVersion}/{build,source}
 
@@ -468,7 +454,7 @@
           cd $dev/lib/modules/${modDirVersion}/source
 
           cp $buildRoot/{.config,Module.symvers} $dev/lib/modules/${modDirVersion}/build
-          make modules_prepare $makeFlags "''${makeFlagsArray[@]}" O=$dev/lib/modules/${modDirVersion}/build
+          make modules_prepare "''${makeFlags[@]}" O=$dev/lib/modules/${modDirVersion}/build
 
           # For reproducibility, removes accidental leftovers from a `cc1` call
           # from a `try-run` call from the Makefile
@@ -584,6 +570,20 @@
 
         makeFlags = [
           "O=$(buildRoot)"
+
+          # We have a `modules` variable in the environment for our
+          # split output, but the kernel Makefiles also define their
+          # own `modules` variable. Their definition wins, but Make
+          # remembers that the variable was originally from the
+          # environment and exports it to all the build recipes. This
+          # breaks the build with an “Argument list too long” error due
+          # to passing the huge list of every module object file in the
+          # environment of every process invoked by every build recipe.
+          #
+          # We use `--eval` here to undefine the inherited environment
+          # variable before any Makefiles are read, ensuring that the
+          # kernel’s definition creates a new, unexported variable.
+          "--eval=undefine modules"
         ] ++ commonMakeFlags;
 
         passthru = { inherit commonMakeFlags; };
```

The reason I turn on `__structuredAttrs` again here is because we need a Make flag with a space in it; without `__structuredAttrs`, we would have to do an escaping dance that would need to be adjusted with `__structuredAttrs`, or else append it in Bash phase code. Since we’re working towards turning on `__structuredAttrs` by default, it’s best to switch it on now and avoid introducing something we’ll have to fix up later. I also dropped `rm -f ${placeholder "modules"}/lib/modules/${modDirVersion}/source` as it doesn’t seem to be necessary any more. (Ideally the `__structuredAttrs` port would be in a separate commit, as before.)

One additional nit is that “nixos/kernel: obtain in-tree kernel modules via the modules output” should come before “linux: install kernel modules to separate output” – `lib.getOutput` will fall back to `out` when the requested output isn’t available, so it’s safe to do it beforehand, and the intermediate state in the current ordering breaks NixOS.

Thanks for working on this and sorry for the back and forth; let me know if you think this diff looks good. I would be happy for you to squash it in with a `Co-authored-by: Emily <vcs@emily.moe>`, or I can force‐push amended commits to your PR branch, or we can discuss it further if you’d like.

jmbaur committed:
nixos/kernel: obtain in-tree kernel modules via the modules output

jmbaur committed:
linux: use __structuredAttrs

Co-authored-by: Emily <vcs@emily.moe>

jmbaur committed:
linux: install kernel modules to separate output

The kernel image is not functionally required at runtime in the same
output where the kernel modules are, and we can save space by removing
it.

Co-authored-by: Emily <vcs@emily.moe>

jmbaur commented:
@emilazy thank you for the patience and recommendations! I've applied your changes, let me know your thoughts

emilazy reviewed (approved):
Looks great, just like I’d have written it 😆

I think this will want to target `master` instead and get manually batched up with the other weekly kernel bump PRs – I’m not a kernel maintainer so I’ll leave that part up to the ones who are.

jmbaur commented:
> I think this will want to target master instead and get manually batched up with the other weekly kernel bump PRs – I’m not a kernel maintainer so I’ll leave that part up to the ones who are.

@k900 any opinions?

jmbaur commented:
Anything else we'd like to see here?

jmbaur commented:
> So this blocks the `nixos-unstable` channel now. Example log: https://hydra.nixos.org/build/305191719/nixlog/1 Example test command:
> 
> ```shell
> nix build -f. nixosTests.allDrivers.initrd-secrets.pigz
> ```

https://github.com/NixOS/nixpkgs/pull/434414

vcunat commented:
Even with that fix this merge broke a channel-blocking build.  It's unclear why, and it happens on aarch64 but not x86_64  🤕   (I was now bisecting with that fix on top)
https://hydra.nixos.org/build/305193018

jmbaur commented:
https://github.com/NixOS/nixpkgs/pull/434968

@vcunat @ElvishJerricco I really appreciate the patience :sweat:

r-vdp commented:
FYI, this broke `nixosTests.systemd-boot.no-bootspec`, since the `synthesize` command of the `bootspec` package expects to find `$toplevel/kernel-modules/bzImage`.
It should probably look at `$toplevel/kernel` instead, but that will require patching bootspec.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/patching-an-in-tree-linux-kernel-module/22137/3
